upgrades.go

```diff
@@ -0,0 +1,186 @@
+package syscfg
+
+// This file contains tweaks for enabling/disabling unattended upgrades.
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"time"
+
+	errw "github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+const (
+	autoUpgradesPath             = "/etc/apt/apt.conf.d/20auto-upgrades"
+	autoUpgradesContentsEnabled  = `APT::Periodic::Update-Package-Lists "1";` + "\n" + `APT::Periodic::Unattended-Upgrade "1";` + "\n"
+	autoUpgradesContentsDisabled = `APT::Periodic::Update-Package-Lists "1";` + "\n" + `APT::Periodic::Unattended-Upgrade "0";` + "\n"
+
+	unattendedUpgradesPath = "/etc/apt/apt.conf.d/50unattended-upgrades"
+)
+
+type UpgradesConfig struct {
+	// Type can be
+	// Empty/missing ("") to make no changes
+	// "disable" (or "disabled") to disable auto-upgrades
+	// "security" to enable ONLY security upgrades
+	// "all" to enable upgrades from all configured sources
+	Type string `json:"type"`
+}
+
+func EnforceUpgrades(ctx context.Context, cfg UpgradesConfig, log *zap.SugaredLogger) {
+	if cfg.Type == "" {
+		return
+	}
+
+	err := checkSupportedDistro()
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	if cfg.Type == "disable" || cfg.Type == "disabled" {
+		isNew, err := writeFileIfNew(autoUpgradesPath, []byte(autoUpgradesContentsDisabled))
+		if err != nil {
+			log.Error(err)
+		}
+		if isNew {
+			log.Info("Disabled OS auto-upgrades.")
+		}
+		return
+	}
+
+	err = verifyInstall()
+	if err != nil {
+		err = doInstall(ctx)
+		if err != nil {
+			log.Error(err)
+			return
+		}
+	}
+
+	securityOnly := cfg.Type == "security"
+	confContents, err := generateOrigins(securityOnly)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	isNew1, err := writeFileIfNew(autoUpgradesPath, []byte(autoUpgradesContentsEnabled))
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	isNew2, err := writeFileIfNew(unattendedUpgradesPath, []byte(confContents))
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	if isNew1 || isNew2 {
+		if securityOnly {
+			log.Info("Enabled OS auto-upgrades (security only.)")
+		} else {
+			log.Info("Enabled OS auto-upgrades (full.)")
+		}
+	}
+
+	err = enableTimer()
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+func checkSupportedDistro() error {
+	data, err := os.ReadFile("/etc/os-release")
+	if err != nil {
+		return err
+	}
+
+	if strings.Contains(string(data), "VERSION_CODENAME=bookworm") || strings.Contains(string(data), "VERSION_CODENAME=bullseye") {
+		return nil
+	}
+
+	return errw.New("cannot enable automatic upgrades for unknown distro, only support for Debian bullseye and bookworm is available")
+}
+
+// make sure the needed package is installed.
+func verifyInstall() error {
+	cmd := exec.Command("unattended-upgrade", "-h")
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return errw.Wrapf(err, "executing 'unattended-upgrade -h' %s", output)
+	}
+	return nil
```

Comment:
this stanza is repeated code; style point so your call, but could be factored out

Reply:
Agree... a bit ugly. Want to reuse this as helper functions across ALL of agent as part of the merge, so didn't factor it out here.